hnn_core/gui/gui.py

```diff
@@ -0,0 +1,1008 @@
+"""IPywidgets GUI."""
+
+# Authors: Mainak Jas <mjas@mgh.harvard.edu>
+#          Huzi Cheng <hzcheng15@icloud.com>
+
+import codecs
+import logging
+import multiprocessing
+import os.path as op
+import sys
+
+import matplotlib.pyplot as plt
+import numpy as np
+
+from IPython.display import display
+from ipywidgets import (HTML, Accordion, AppLayout, BoundedFloatText,
+                        BoundedIntText, Button, Dropdown, FileUpload,
+                        FloatLogSlider, FloatText, GridspecLayout, HBox,
+                        IntText, Label, Layout, Output, RadioButtons, Tab,
+                        Text, VBox, interactive_output)
+
+import hnn_core
+from hnn_core import (JoblibBackend, MPIBackend, jones_2009_model, read_params,
+                      simulate_dipole)
+from hnn_core.params import (_extract_drive_specs_from_hnn_params, _read_json,
+                             _read_legacy_params)
+from hnn_core.viz import plot_dipole
+
+
+THEMECOLOR = "#8A2BE2"
+
+
+def create_expanded_button(description, button_style, height, disabled=False):
+    style = {'button_color': THEMECOLOR}
+    return Button(description=description,
+                  button_style=button_style,
+                  layout=Layout(height=height, width='auto'),
+                  style=style,
+                  disabled=disabled)
+
+
+def _get_sliders(params, param_keys):
+    """Get sliders"""
+    style = {'description_width': '150px'}
+    sliders = list()
+    for d in param_keys:
+        slider = FloatLogSlider(value=params[d],
+                                min=-5,
+                                max=1,
+                                step=0.2,
+                                description=d.split('gbar_')[1],
+                                disabled=False,
+                                continuous_update=False,
+                                orientation='horizontal',
+                                readout=True,
+                                readout_format='.2e',
+                                style=style)
+        sliders.append(slider)
+
+    def _update_params(**updates):
+        logging.debug(f'Connectivity parameters updates: {updates}')
+        params.update(dict(**updates))
+
+    interactive_output(_update_params, {s.description: s for s in sliders})
+    return sliders
+
+
+def _get_cell_specific_widgets(layout, style, location, data=None):
+    default_data = {
+        'weights_ampa': {
+            'L5_pyramidal': 0.,
+            'L2_pyramidal': 0.,
+            'L5_basket': 0.,
+            'L2_basket': 0.
+        },
+        'weights_nmda': {
+            'L5_pyramidal': 0.,
+            'L2_pyramidal': 0.,
+            'L5_basket': 0.,
+            'L2_basket': 0.
+        },
+        'delays': {
+            'L5_pyramidal': 0.1,
+            'L2_pyramidal': 0.1,
+            'L5_basket': 0.1,
+            'L2_basket': 0.1
+        },
+    }
```

Comment:
This is perhaps opening a large can of worms, but what happens when we want to add an hnn-core model that contain more or less cell types than the ones we currently have? Will we need to update the hardcoded cell types here?

Reply:
yep. But the widgets will also need to be updated ... no point in generalizing just this.

Reply:
Unless I'm missing something crucial here, I feel like the GUI code should be as independent from specific network attributes as possible. If we hardcode these values in now, we're just repeating the same mistakes of the past GUI.

As a solution, I propose that we first have the user specify which network model to load into the GUI environment, and then instantiate cell specific widgets based on the cell types and properties contained in that `Network` instance.